data-prepper-plugins/saas-source-plugins/crowdstrike-source/src/main/java/org/opensearch/dataprepper/plugins/source/crowdstrike/CrowdStrikeService.java

```diff
@@ -0,0 +1,87 @@
+import static org.opensearch.dataprepper.plugins.source.crowdstrike.utils.Constants.FILTER_KEY;
+import static org.opensearch.dataprepper.plugins.source.crowdstrike.utils.Constants.LAST_UPDATED;
+import static org.opensearch.dataprepper.plugins.source.crowdstrike.utils.Constants.LIMIT_KEY;
+
+/**
+ * This service manages the interaction with CrowdStrike's API endpoints, handles pagination
+ * and provides methods for fetching indicators from CrowdStrike.
+ */
+@Named
+public class CrowdStrikeService {
+
+    private final CrowdStrikeRestClient crowdStrikeRestClient;
+    private static final Logger log = LoggerFactory.getLogger(CrowdStrikeService.class);
+    private static final String BASE_URL = "https://api.crowdstrike.com/";
+    private static final String COMBINED_URL = "https://api.crowdstrike.com/intel/combined/indicators/v1";
+    private final Timer searchCallLatencyTimer;
+
+
+    public CrowdStrikeService(CrowdStrikeRestClient crowdStrikeRestClient, PluginMetrics pluginMetrics) {
+        this.crowdStrikeRestClient = crowdStrikeRestClient;
+        this.searchCallLatencyTimer = pluginMetrics.timer("searchCallLatencyTimer");
+    }
+
+    /**
+     * Retrieves all indicator data from CrowdStrike in a paginated fashion.
+     *   @param startTime       The start timestamp (inclusive) for filtering indicators.
+     *   @param endTime         The end timestamp (exclusive) for filtering indicators.
+     *   @param paginationLink  An optional pagination URL suffix (used when fetching next pages).
+     *   @return                A {@link CrowdStrikeApiResponse} containing response body and headers.
+     */
+    public CrowdStrikeApiResponse getThreatIndicators(Instant startTime, Instant endTime, Optional<String> paginationLink) {
+        if (startTime == null || endTime == null) {
+            throw new IllegalArgumentException("startTime and endTime must not be null");
+        }
+        URI uri = buildCrowdStrikeUri(startTime, endTime, paginationLink);
+        return searchCallLatencyTimer.record(() -> {
+
+            log.debug("Calling CrowdStrike API with URI: {}", uri);
+            ResponseEntity<CrowdStrikeIndicatorResult> responseEntity = crowdStrikeRestClient.invokeGetApi(uri, CrowdStrikeIndicatorResult.class);
+
+            return new CrowdStrikeApiResponse(responseEntity.getBody(), responseEntity.getHeaders());
+        });
+    }
+
+    protected URI buildCrowdStrikeUri(Instant startTime, Instant endTime,  Optional<String> paginationLink) {
+        try {
+            if (paginationLink.isPresent()) {
+                String urlString = BASE_URL + paginationLink.get();
+                urlString = CrowdStrikeNextLinkValidator.validateAndSanitizeURL(urlString);
+                return new URI(urlString);
+            } else {
+                // Manually construct and encode the query string
+                String filter1 = URLEncoder.encode(LAST_UPDATED + ":>=" + startTime.getEpochSecond(), StandardCharsets.UTF_8);
```

Comment:
minor: plz also use CONSTANTS like GREATER_THAN, LESSER_THAN and explain in code comments with an example of what the final url will look like for readability